[fix][schema] Reject unsupported Avro schema types during schema registration

poorbarcode opened the pull request:
### Motivation

#### Issue
- User can upload an Avro-Union schema by `pulsar-admin` API
- However, the schema uploaded can not be used, users will get an error "Not a record" when creating a producer with `AUTO_PRODUCE` schema.

### Modifications
- Deny uploading an Avro-Union schema
- Improve the logs when creating a producer failed due to an invalidated schema

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: x

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/SchemasResourceBase.java:
```diff
@@ -123,8 +125,28 @@ public CompletableFuture<SchemaVersion> deleteSchemaAsync(boolean authoritative,
+        switch (SchemaType.valueOf(payload.getType())) {
+            case AVRO : {
+                Schema schema = SchemaUtil.parseAvroSchema(payload.getSchema());
+                try {
+                    // Pulsar only support "RecordSchema" so far. Other types will throw an error when calls
+                    // "getFields".
+                    // Since "RecordSchema" is an avro internal private class, we can not use "instanceof" here.
+                    schema.getFields();
+                } catch (Exception e) {
+                    return CompletableFuture.failedFuture(new RestException(Response.Status.BAD_REQUEST.getStatusCode(),
+                          "[" + String.valueOf(topicName) + "] Avro schema typed [" + schema.getType() + "]"
+                          + " is not supported"));
+                }
```
There's also a public `org.apache.avro.Schema#isUnion` method if that's the only case to check.

poorbarcode replied:
Thanks, improved. And added a new test

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/SchemasResourceBase.java:
```diff
@@ -123,8 +125,41 @@ public CompletableFuture<SchemaVersion> deleteSchemaAsync(boolean authoritative,
+                                Response.Status.BAD_REQUEST.getStatusCode(),
+                                "[" + String.valueOf(topicName) + "] Avro schema typed [UNION] is not supported"));
+                    }
+                    case INT:
+                    case LONG:
+                    case FLOAT:
+                    case DOUBLE:
+                    case BOOLEAN:
+                    case STRING:
+                    case BYTES: {
```
would it make sense to have a default case instead?

poorbarcode replied:
> would it make sense to have a default case instead?

No, if users can register an Avro schema successfully, they will try to use the same schema to consume, and then they will get an error. We'd better deny this operation

lhotari replied:
> > would it make sense to have a default case instead?
> 
> No, if users can register an Avro schema successfully, they will try to use the same schema to consume, and then they will get an error. We'd better deny this operation

@poorbarcode what I meant is using a `default` case for the `switch`, instead of listing all other cases. Does your comment apply to that?

lhotari replied:
@poorbarcode Thanks for checking. Yes, the current approach makes sense. I'll mark this resolved.

lhotari reviewed (approved):
LGTM, good work @poorbarcode

poorbarcode commented:
/pulsarbot rerun-failure-checks

poorbarcode commented:
/pulsarbot rerun-failure-checks

lhotari commented:
There's a test failure:
```
  Error:  Tests run: 175, Failures: 1, Errors: 0, Skipped: 165, Time elapsed: 46.633 s <<< FAILURE! - in org.apache.pulsar.broker.admin.AdminApiSchemaTest
  Error:  org.apache.pulsar.broker.admin.AdminApiSchemaTest.testInvalidSchemaDataException[V1](4)  Time elapsed: 0.01 s  <<< FAILURE!
  java.lang.AssertionError: expected [422] but found [500]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1577)
  	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
  	at org.testng.Assert.assertEquals(Assert.java:131)
  	at org.testng.Assert.assertEquals(Assert.java:1418)
  	at org.testng.Assert.assertEquals(Assert.java:1382)
  	at org.testng.Assert.assertEquals(Assert.java:1428)
  	at org.apache.pulsar.broker.admin.AdminApiSchemaTest.testInvalidSchemaDataException(AdminApiSchemaTest.java:282)
  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  	at java.base/java.lang.Thread.run(Thread.java:1583)
```

lhotari commented:
@poorbarcode It seems that this test failure remainns:
```
  Error:  Tests run: 41, Failures: 1, Errors: 0, Skipped: 17, Time elapsed: 151.005 s <<< FAILURE! - in org.apache.pulsar.schema.SchemaTest
  Error:  org.apache.pulsar.schema.SchemaTest.testSendAvroAndJsonPrimitiveSchema  Time elapsed: 0.187 s  <<< FAILURE!
  org.apache.pulsar.client.api.PulsarClientException$IncompatibleSchemaException: {"errorMsg":"Please call org.apache.pulsar.client.api.Schema.INT when using a simple type schema","reqId":3649047843048571735, "remote":"localhost/127.0.0.1:45669", "local":"/127.0.0.1:38284"}
  	at org.apache.pulsar.client.api.PulsarClientException.unwrap(PulsarClientException.java:1067)
  	at org.apache.pulsar.client.impl.TypedMessageBuilderImpl.send(TypedMessageBuilderImpl.java:107)
  	at org.apache.pulsar.schema.SchemaTest.testSendAvroAndJsonPrimitiveSchema(SchemaTest.java:408)
  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  	at java.base/java.lang.Thread.run(Thread.java:1583)
```

poorbarcode commented:
> @poorbarcode It seems that this test failure remainns:

Sorry, the schema typed `Avro & Primitive` is supported, I have removed the check of these schemas and added a new test to ensure no break changes in the future: `testAvroIntSchema`
